src/plugins/action/action_impl.cpp

```diff
@@ -389,18 +389,34 @@ void ActionImpl::goto_location_async(
     const float yaw_deg,
     const Action::ResultCallback& callback)
 {
-    MavlinkCommandSender::CommandInt command{};
-
-    command.command = MAV_CMD_DO_REPOSITION;
-    command.target_component_id = _parent->get_autopilot_id();
-    command.params.param4 = to_rad_from_deg(yaw_deg);
-    command.params.x = int32_t(std::round(latitude_deg * 1e7));
-    command.params.y = int32_t(std::round(longitude_deg * 1e7));
-    command.params.z = altitude_amsl_m;
-
-    _parent->send_command_async(
-        command, [this, callback](MavlinkCommandSender::Result result, float) {
-            command_result_callback(result, callback);
+    // Change to Hold mode first
+    _parent->set_flight_mode_async(
+        SystemImpl::FlightMode::Hold,
+        [this, yaw_deg, latitude_deg, longitude_deg, altitude_amsl_m, callback](
+            MavlinkCommandSender::Result mode_change_result, float) {
+            Action::Result action_result = action_result_from_command_result(mode_change_result);
+
+            // If successfully changed to Hold, then send the DO_REPOSITION command
+            if (action_result == Action::Result::Success) {
+                MavlinkCommandSender::CommandInt command{};
+
+                command.command = MAV_CMD_DO_REPOSITION;
+                command.target_component_id = _parent->get_autopilot_id();
+                command.params.param4 = to_rad_from_deg(yaw_deg);
+                command.params.x = int32_t(std::round(latitude_deg * 1e7));
+                command.params.y = int32_t(std::round(longitude_deg * 1e7));
+                command.params.z = altitude_amsl_m;
+
+                _parent->send_command_async(
+                    command,
+                    [this, callback](MavlinkCommandSender::Result do_reposition_result, float) {
+                        command_result_callback(do_reposition_result, callback);
+                    });
+            } else {
+                if (callback) {
+                    callback(action_result);
```

Comment:
Same as above:

```
_parent->send_command_async(...
```

Reply:
I don't get what you mean

Reply:
```suggestion
                    _parent->send_command_async(
                        command,
                        [this, callback](MavlinkCommandSender::Result mode_change_result, float) {
                            command_result_callback(mode_change_result_result, callback);
                        });
```

I mean something like this, where you put the callback on the queue instead of calling it directly.

But I haven't tested it :innocent:

Reply:
Ah got it

Reply:
actually I am still trying to see where that is the correct approach. I just followed for example what happens with `arm_async` and it looks similar to what I did. I ended up mimic the approach there, as you can see on the commit now. And I guess that's the correct approach considering it's also being applied to the arm command.